File: oauth-authenticator-archetype/src/main/resources/archetype-resources/src/main/java/authentication/__pluginName__AuthenticatorRequestHandler.java
package ${package}.authentication;

import ${package}.config.${pluginName}AuthenticatorPluginConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import se.curity.identityserver.sdk.authentication.AuthenticationResult;
import se.curity.identityserver.sdk.authentication.AuthenticatorRequestHandler;
import se.curity.identityserver.sdk.service.ExceptionFactory;
import se.curity.identityserver.sdk.service.Json;
import se.curity.identityserver.sdk.service.authentication.AuthenticatorInformationProvider;
import se.curity.identityserver.sdk.web.Request;
import se.curity.identityserver.sdk.web.Response;

import java.util.Optional;

public class ${pluginName}AuthenticatorRequestHandler implements AuthenticatorRequestHandler<RequestModel> {
    private static final Logger _logger = LoggerFactory.getLogger(${pluginName}AuthenticatorRequestHandler.class);

    private final ${pluginName}AuthenticatorPluginConfig _config;
    private final AuthenticatorInformationProvider _authInfoProvider;
    private final ExceptionFactory _exceptionFactory;

    public ${pluginName}AuthenticatorRequestHandler(${pluginName}AuthenticatorPluginConfig config,
                                                    ExceptionFactory exceptionFactory,
                                                    Json json,
                                                    AuthenticatorInformationProvider provider) {
        _config = config;
        _oauthClient = new DefaultOAuthClient(exceptionFactory, provider, json, config.getSessionManager());
    }

    @Override
    public Optional<AuthenticationResult> get(RequestModel requestModel, Response response) {
        _logger.info("GET request received for authentication");

        _oauthClient.setServiceProviderId(requestModel.getRequest());

        _oauthClient.redirectToAuthorizationEndpoint(response,
                _config.getAuthorizationEndpoint().toString(),
                _config.getClientId(),
                _config.getScope(), ImmutableMap.of(PARAM_REDIRECT_URI, _oauthClient.getCallbackUrl());

        return Optional.empty();
    }

    @Override
    public Optional<AuthenticationResult> post(RequestModel requestModel, Response response) {
        return Optional.empty();
    }

    @Override
    public RequestModel preProcess(Request request, Response response) {
        return new RequestModel(request);
    }
}

File: kotlin-oauth-authenticator-archetype/src/main/resources/archetype-resources/src/main/kotlin/authentication/__pluginName__AuthenticatorRequestHandler.kt
package ${package}.authentication

import ${package}.config.${pluginName}AuthenticatorPluginConfig
import ${package}.authentication.RedirectUriUtil.Companion.createRedirectUri
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import se.curity.identityserver.sdk.attribute.Attribute
import se.curity.identityserver.sdk.authentication.AuthenticationResult
import se.curity.identityserver.sdk.authentication.AuthenticatorRequestHandler
import se.curity.identityserver.sdk.errors.ErrorCode
import se.curity.identityserver.sdk.http.RedirectStatusCode
import se.curity.identityserver.sdk.service.ExceptionFactory
import se.curity.identityserver.sdk.service.authentication.AuthenticatorInformationProvider
import se.curity.identityserver.sdk.web.Request
import se.curity.identityserver.sdk.web.Response

import java.net.MalformedURLException
import java.net.URL
import java.util.Optional
import java.util.UUID

class ${pluginName}AuthenticatorRequestHandler(private val _config: ${pluginName}AuthenticatorPluginConfig): AuthenticatorRequestHandler<Request>
{
    private val _authenticatorInformationProvider: AuthenticatorInformationProvider = _config.getAuthenticatorInformationProvider()
    private val _exceptionFactory: ExceptionFactory = _config.getExceptionFactory()

    override fun get(request: Request, response: Response): Optional<AuthenticationResult>
    {
        _logger.debug("GET request received for authentication")

        val redirectUri = createRedirectUri(_authenticatorInformationProvider, _exceptionFactory)
        val state = UUID.randomUUID().toString()
        val scopes = setOf("scope")

        _config.getSessionManager().put(Attribute.of("state", state))

        val queryStringArguments = linkedMapOf<String, Collection<String>>(
                "client_id" to setOf(_config.getClientId()),
                "redirect_uri" to setOf(redirectUri),
                "state" to setOf(state),
                "response_type" to setOf("code"),
                "scope" to setOf(scopes.joinToString(" "))
        )

        _logger.debug("Redirecting to {} with query string arguments {}", AUTHORIZATION_ENDPOINT,
                queryStringArguments)

        throw _exceptionFactory.redirectException(AUTHORIZATION_ENDPOINT,
                RedirectStatusCode.MOVED_TEMPORARILY, queryStringArguments, false)
    }

    override fun post(request: Request, response: Response): Optional<AuthenticationResult>
    {
        throw _exceptionFactory.methodNotAllowed()
    }

    override fun preProcess(request: Request, response: Response): Request
    {
        return request
    }

    companion object
    {
        private val _logger: Logger = LoggerFactory.getLogger(${pluginName}AuthenticatorRequestHandler::class.java)
        private const val AUTHORIZATION_ENDPOINT = ""

    }
}
